osu.Game/Rulesets/Ruleset.cs

```diff
@@ -389,5 +389,13 @@ public IEnumerable<(HitResult result, LocalisableString displayName)> GetHitResu
         /// Can be overridden to alter the difficulty section to the editor beatmap setup screen.
         /// </summary>
         public virtual DifficultySection? CreateEditorDifficultySection() => null;
+
+        public virtual double PreemptFromAr(float AR) => 0;
+        public virtual float ArFromPreempt(double preempt) => 5;
+        public virtual double HitwindowFromOd(float OD) => 0;
+        public virtual float OdFromHitwindow(double hitwindow300) => 0;
+
+        public virtual float ChangeArFromRate(float AR, double rate) => AR;
+        public virtual float ChangeOdFromRate(float OD, double rate) => OD;
```

Comment:
No. This is not acceptable. What happened with [yesterday's proposal](https://discord.com/channels/188630481301012481/1097318920991559880/1144338816400900198) of having `public virtual BeatmapDifficulty GetEffectiveDifficulty()`?

To explain why this is not acceptable: rulesets can be and do all sorts of things. How do you define this if a ruleset has no concept of "approach rate"? How do you define this if a ruleset has no concept of "hit windows" (catch)?

Reply:
This is made for conveniency. I thought it will be useful to use one function instead of using inline implementations of "AR to preempt" kinda things. 
The main functions here are `ChangeArFromRate` and `ChangeOdFromRate`.

About the `GetEffectiveDifficulty()`.
```
isArRateAdjusted = false;
isOdRateAdjusted = false;
```
Those two lines are made to indicate that AR/OD changed thru rate (and not "true"). This kinda goes against the idea of `GetEffectiveDifficulty()` that just returns effective difficulty.